internal/cloudapi/server.go

```diff
@@ -182,7 +182,7 @@ func (server *Server) Compose(w http.ResponseWriter, r *http.Request) {
 		for name, packages := range packageSets {
 			pkgs, _, err := server.rpmMetadata.Depsolve(packages, repositories, distribution.ModulePlatformID(), arch.Name())
 			if err != nil {
-				http.Error(w, fmt.Sprintf("Failed to depsolve base packages for %s/%s/%s: %s", ir.ImageType, ir.Architecture, request.Distribution, err), http.StatusInternalServerError)
+				http.Error(w, fmt.Sprintf("Failed to depsolve base packages for %s/%s/%s: %s", ir.ImageType, ir.Architecture, request.Distribution, err), http.StatusBadRequest)
```

Comment:
I wonder, could we distinguish between the different error types? I.e., distinguish between "conflicting packages"/"unknown package" (should return 400) on the one hand, and "file not found"/<genuine internal error> on the other (should be 500)? In the case of a 400 it would be ideal to return some structured error that we could eventually expose in some nice way to the user. I.e., an array of packages that were not found, or some structured information about conflicting requests.

The structured errors should probably be done in a follow-up PR, distinguishing between the error codes could be done here or in a follow-up, depending on how much work that would be, this is definitely already an improvement not he existing behaviour.

Reply:
Thanks for the comment!

I agree with you. And it seems that there is an issue with my test case anyway.
I've got a weird error `Failed to depsolve base packages for ami/x86_64/rhel-8: unexpected end of JSON input`

Looking at my logs, I can see that the parsing of my input json fails.

```
Jul 16 16:05:50 fedora python3[682563]: detected unhandled Python exception in '/usr/libexec/osbuild-composer/dnf-json'
Jul 16 16:05:50 fedora osbuild-composer[682563]: Traceback (most recent call last):
Jul 16 16:05:50 fedora osbuild-composer[682563]:   File "/usr/libexec/osbuild-composer/dnf-json", line 117, in <module>
Jul 16 16:05:50 fedora osbuild-composer[682563]:     base = create_base(
Jul 16 16:05:50 fedora osbuild-composer[682563]:   File "/usr/libexec/osbuild-composer/dnf-json", line 70, in create_base
Jul 16 16:05:50 fedora osbuild-composer[682563]:     for repo in repos:
Jul 16 16:05:50 fedora osbuild-composer[682563]: TypeError: 'NoneType' object is not iterable
```

Something is off with either my test scenario or something else and I need to investigate.

Either way, in this specific configuration, the result should clearly be a 500 error and not a 400 as the server crashed.

If it is right for you, I'd like to explore answering the question within this PR as I think my actual solution is too simple to be considered the right answer. And correcting it may result in a PR with the feature you were talking about.

Reply:
Ah, indeed, an exception should be a 500. Good catch!

The problem is that your request does not contain any repositories, but we should either check for this and return 400, or we should handle it gracefully in `dns-json`.

> If it is right for you, I'd like to explore answering the question within this PR as I think my actual solution is too simple to be considered the right answer.

Sounds perfect!